readthedocs/proxito/views/serve.py

```diff
@@ -274,6 +286,172 @@ def _get_canonical_redirect_type(self, request):
+            )
+
+        project = unresolved.project
+        version = unresolved.version
+        filename = unresolved.filename
+
+        # TODO: find a better way to pass this to the middleware.
+        request.path_project_slug = project.slug
+        request.path_version_slug = version.slug
+
+        if not version.active:
+            log.warning("Version is not active.")
+            raise Http404("Version is not active.")
+
+        if self._is_cache_enabled(project) and not version.is_private:
+            # All public versions can be cached.
+            self.cache_request = True
+
+        log.bind(cache_request=self.cache_request)
+        log.debug("Serving docs.")
+
+        # Verify if the project is marked as spam and return a 401 in that case
+        spam_response = self._spam_response(request, project)
+        if spam_response:
+            return spam_response
+
+        # Check for canonical redirects before serving the file.
+        redirect_type = self._get_canonical_redirect_type(request)
+        if redirect_type:
+            try:
+                return self.canonical_redirect(
+                    request=request,
+                    final_project=project,
+                    version_slug=version.slug,
+                    filename=filename,
+                    redirect_type=redirect_type,
+                    is_external_version=unresolved.external,
+                )
+            except InfiniteRedirectException:
+                # Don't redirect in this case, since it would break things.
+                pass
+
+        # Trailing slash redirect.
+        # We don't want to serve documentation at:
+        # - `/en/latest`
+        # - `/projects/subproject/en/latest`
+        # - `/projects/subproject`
+        # These paths need to end with an slash.
+        if filename == "/" and not path.endswith("/"):
+            return self.system_redirect(
+                request=request,
+                final_project=project,
+                version_slug=version.slug,
+                filename=filename,
+                is_external_version=unresolved_domain.is_from_external_domain,
+            )
+
+        # Check for forced redirects.
+        redirect_path, http_status = self.get_redirect(
+            project=project,
+            lang_slug=project.language,
+            version_slug=version.slug,
+            filename=unresolved.filename,
+            full_path=request.path,
+            forced_only=True,
+        )
+        if redirect_path and http_status:
+            log.bind(forced_redirect=True)
+            try:
+                return self.get_redirect_response(
+                    request=request,
+                    redirect_path=redirect_path,
+                    proxito_path=request.path,
+                    http_status=http_status,
+                )
+            except InfiniteRedirectException:
+                # Continue with our normal serve.
+                pass
```

Comment:
We should log a warning here.

Reply:
We already have a warning in the function itself https://github.com/readthedocs/readthedocs.org/blob/d499203f6edf0106f146941ac6c120d831e916e8/readthedocs/proxito/views/mixins.py#L456-L459

Reply:
Having these bubble up and be ignored in multiple places definitely feels odd. Should we actually be exposign the error to the user, since it's some kind of misconfiguration of redirects?

Reply:
That's probably a better outcome actually. Is there a specific HTTP code response for this?